### PR TITLE
Add separate etcd client keys for calico and flanneld

### PR DESCRIPTION
**Pull request by tuommaki**
So far all etcd clients used server keys which is not secure
authentication wise. Add separate etcd client keys for flanneld and
calico.

Towards https://github.com/giantswarm/giantswarm/issues/1500 and https://github.com/giantswarm/giantswarm/issues/1559

### CODE REVIEW

File: types.go
Comment (tuommaki) [inline types.go]: Sorted alphabetically while here.

File: types.go
Comment (tuommaki) [inline types.go]: Sorted alphabetically while added Calico&Flanneld Etcd Client entries.

**Review by xh3b4sd — APPROVED** [review]

**Review by r7vme — APPROVED** [review]
LGTMIIWFY

**Review by kopiczko — APPROVED** [review]

**Comment by kopiczko** [comment]
Note that operator updates may be tricky here. Do you have plan for the migration?

**Comment by tuommaki** [comment]
My thinking was to roll this out when `cluster-operator` starts managing `CertConfigs`, but there's the point that this could conflict with other changes to here. I'll put back the temporary client key as well and let's remove it when `cluster-operator` rollout has happened for this.

**Comment by tuommaki** [comment]
> The problem is changing FS paths guest clusters somehow rely on. We cannot migrate this with cluster-operator alone. I missed this point in my review.

Ah, you are right. I think this requires yet another `hive` / `installations` / other repos "cert aligning" trip. Nevertheless, in current state, this should not break existing clusters and this should create a path towards client specific keys so that there can be progress in other components, right?

**Comment by xh3b4sd** [comment]
Current state of the diff looks good again.

**Comment by kopiczko** [comment]
So existing cloud configs are generated. On upgrade the new ones with new paths should be generated - machines are replaced. I'm not sure if we don't use the certs elsewhere dynamically. But you'd have to be 100% sure. I feel like it should work event w/o temporary paths, but would need to be tested.

File: files.go
Comment (kopiczko) [inline files.go]: Maybe it would be good to put current operator versions here, so we know when we are safe to delete it. WDYT?

**Comment by xh3b4sd** [comment]
When we only add new key pairs here nothing can go wrong. The migration path is split into different stages of state transitions. Once we have all key pairs in place we can use them where ever we want. Later we can drop what we do not need anymore. All good.

File: files.go
Comment (xh3b4sd) [reply files.go]: Interesting idea but I do not follow how you want this to be. How should this look like?

Comment (kopiczko) [reply files.go]: Like

```
Version bundle versions at the time of writing this comment:

    - flannel-operator: v...
    - azure-operator: v0.1.0
    - ...
```

Then we know when those versions bundles (and all lower versions) are removed we are safe to remove migration code.

**Review by calvix — CHANGES_REQUESTED** [review]
I am a little bit lost. How does this work?
I mean in general. I have never seen this lib and description is quite vague.

We don't use cert-operator for certificates for host cluster so what is flannel doing there?
atm we mount etcd files from machine

I don't see how could cert-operator create certs on host cluster as it runs as a single pod on a random machine.

Also, flanneld and rest of certs are in different group as flanneld needs certs for host cluster ETCD and rest need certs for k8s/etcd for the guest cluster which are different CA and PKI.

We could drop flannel certs on host machines if we properly created them as secrets and then properly mounted into pods. But there has to be lot of changes and also we need to be sure certs are ready before flannel pod is started.

**Comment by kopiczko** [comment]
@calvix this is used only by operators to create guest clusters.

File: files.go
Comment (xh3b4sd) [reply files.go]: Versioning in comments will not work.

Comment (kopiczko) [reply files.go]: How can you know when you can remove this code otherwise?

Comment (xh3b4sd) [reply files.go]: It is ok now Pawel. Do what you think is right. Disagree and commit. I do not have strong feelings here and no further motivation to get deeper into the topic right now because my focus is actually somewhere else and I want to keep it like this. Did not want to stress this here.

**Comment by calvix** [comment]
Then flannel certificates should not be there as that is not relevant.

File: files.go
Comment (kopiczko) [reply files.go]: It isn't my PR :D 

I'm also fine with not stating current versions. But I just wonder how to delete such (at some point) dead code.

**Comment by kopiczko** [comment]
@calvix this is etcd certificate which is used to store flannel configuration.

**Comment by xh3b4sd** [comment]
The flannel-operator configures certs for the flannelds it is launching via daemon sets. Right now we reuse the operator's certs for the guest clusters. We want to decouple this. See also https://github.com/giantswarm/flannel-operator/blob/8b51e434e43ac8a9bb00965a66e5765c537107f4/service/flannelconfig/v3/resource/legacy/daemonset.go#L72-L83.

**Comment by calvix** [comment]
> The flannel-operator configures certs for the flannelds it is launching via daemon sets. Right now we reuse the operator's certs for the guest clusters. We want to decouple this.

I am pretty sure we don't reuse guest cluster certs, as they have completely different PKI backend than flanneld needs.
for flanneld we use certs that are mounted from disk from host cluster. Flanneld needs access to host cluster etcd.

**Comment by xh3b4sd** [comment]
We reuse the operator's certs, not the guest cluster ones. Further, is there anything why you think we should not merge this? If not please approve the PR or elaborate on your concerns to we can move forward.

**Review by calvix — APPROVED** [review]
From my point that flannel part in this part is nonsense, I already said my points. Go as you wish

**Comment by xh3b4sd** [comment]
It is correct because we use the certs we mount for the flannel-operator for the guest clusters as well. With this we will be able to not reuse the flannel-operator's certs in the guest clusters and give the guest cluster's flannelds separate key pairs.

**Comment by tuommaki** [comment]
Ok. Studied how stuff works on KVM and had a discussion with @calvix and end result is that flannel keys are useless here because they are mounted from the host machine so ones created by `cert-operator` wouldn't get picked by anything. Calico keys are still useful so I'll keep them & merge.

**Comment by xh3b4sd** [comment]
For the record, the important bit we missed here was that the flannelds run for the guest clusters are not talking to the guest cluster etcd, but to the host cluster etcd, which makes guest cluster certs useless, which is why we use the same host cluster certs. Anyway, we could still issue separate certs here because the host cluster uses Vault the same way. The only difference is the used PKI backend in Vault which is identified by a cluster ID. So the mechanism should be the same, only the config should be slightly different. @giantswarm/sig-security what is your thinking here? Important? Should we do this? @tuommaki I think this should not be too much different from your proposal here. @calvix this should then work right?

**Comment by puja108** [comment]
We have a story about separating certs for components in guest clusters, but I think we should also do the same in host cluster, just not sure if we should do it "on the side" or if this should be a concentrated effort. My thinking: if we can just get it done for this now without too much additional work (boyscouting timebox) let's go, otherwise let's do a general security rework of our certs separating each one of them, so we do not reuse at all.

**Comment by calvix** [comment]
@xh3b4sd yes, then we need to update flannel dand mount the certs to pod from secrets
in genral this give us more flexibility when we need to change stuff in the certs

**Pull request merged**
